ramips: add support for TP-Link RE500 v1

achterin opened the pull request:
This device uses the same hardware as RE650 v1 which got supported in
8c51dde.

Hardware specification:

- SoC 880 MHz - MediaTek MT7621AT
- 128 MB of DDR3 RAM
- 16 MB - Winbond 25Q128FVSG
- 4T4R 2.4 GHz - MediaTek MT7615E
- 4T4R 5 GHz - MediaTek MT7615E
- 1x 1 Gbps Ethernet - MT7621AT integrated
- 7x LEDs (Power, 2G, 5G, WPS(x2), Lan(x2))
- 4x buttons (Reset, Power, WPS, LED)
- UART header (J1) - 2:GND, 3:RX, 4:TX
  Serial console @ 57600,8n1

Flash instructions:

Upload
openwrt-ramips-mt7621-tplink_re500-v1-squashfs-factory.bin
from the RE500 web interface.

TFTP recovery to stock firmware:

Unfortunately, I can't find an easy way to recover the RE
without opening the device and using modified binaries. The
TFTP upload will only work if selected from u-boot, which
means you have to open the device and attach to the serial
console. The TFTP update procedure does *not* accept the
published vendor firmware binaries. However, it allows to
flash kernel + rootfs binaries, and this works if you have
a backup of the original contents of the flash. It's probably
possible to create special image out of the vendor binaries
and use that as recovery image.

adrianschmutzler commented:
Since this is for mt7621, I intend to wait until kernel 5.4 support is added, in order to not create rebase-hell for the people working on it.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_tplink_re500-v1.dts:
```diff
@@ -0,0 +1,48 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_power: power {
```
Since LEDs are defined in the DTS files, please also move the led-aliases there; label-mac-device may stay in the DTSI.

achterin replied:
done

adrianschmutzler commented:
@achterin I've updated this to work with kernel 5.4 (supposedly) and pushed it to my staging tree. Please have a look whether everything is still correct.

Despite, we should convert the WiFi LEDs to use DTS triggers. Can you derive the correct values and report them here, so I can update the patch.

@gvlaev You did support for RE650 v1, I'd be happy if we could transfer that one to DTS triggers as well.

gvlaev commented:
@adrianschmutzler Sure, I'll add the triggers.

achterin commented:
> Despite, we should convert the WiFi LEDs to use DTS triggers. Can you derive the correct values and report them here, so I can update the patch.

i tried that approach before submitting the pull request in the first place but found out, that there were no throughput triggers present. i need to check some other mt7615 single band device whether or not the trigger is present. if so, there might be an issue with multi phy handling in the driver.

adrianschmutzler commented:
@gvlaev @achterin Ping. Note that mt76 has been updated yesterday ...

achterin commented:
@adrianschmutzler thanks for pinging again. sadly my device is at the hackerspace and i cant get it right now as we dont meet in person because of corona. i will update the pr and retest everything once i get my hands on the device again. you can add the blocker tag for now if you want. sorry for the delay.

adrianschmutzler commented:
@achterin That's okay, let's just wait ...

adrianschmutzler commented:
Just another ping. If the situation hasn't changed, just ignore it.

achterin committed:
ramips: add support for TP-Link RE500 v1

This device uses the same hardware as RE650 v1 which got supported in
8c51dde.

Hardware specification:

- SoC 880 MHz - MediaTek MT7621AT
- 128 MB of DDR3 RAM
- 16 MB - Winbond 25Q128FVSG
- 4T4R 2.4 GHz - MediaTek MT7615E
- 4T4R 5 GHz - MediaTek MT7615E
- 1x 1 Gbps Ethernet - MT7621AT integrated
- 7x LEDs (Power, 2G, 5G, WPS(x2), Lan(x2))
- 4x buttons (Reset, Power, WPS, LED)
- UART header (J1) - 2:GND, 3:RX, 4:TX
  Serial console @ 57600,8n1

Flash instructions:

Upload
openwrt-ramips-mt7621-tplink_re500-v1-squashfs-factory.bin
from the RE500 web interface.

TFTP recovery to stock firmware:

Unfortunately, I can't find an easy way to recover the RE
without opening the device and using modified binaries. The
TFTP upload will only work if selected from u-boot, which
means you have to open the device and attach to the serial
console. The TFTP update procedure does *not* accept the
published vendor firmware binaries. However, it allows to
flash kernel + rootfs binaries, and this works if you have
a backup of the original contents of the flash. It's probably
possible to create special image out of the vendor binaries
and use that as recovery image.

Signed-off-by: Christoph Krapp <achterin@googlemail.com>

achterin commented:
@adrianschmutzler rebased my changes on the current master. i've retried the dts-trigger approach which worked this time. as i dont have a re350-v1 i did split the led setup code.

adrianschmutzler commented:
> as i dont have a re350-v1 i did split the led setup code.

Err, the re350-v1 is completely separate?!

>  rebased my changes on the current master. i've retried the dts-trigger approach which worked this time.

Thanks, I've taken the patch with a few adjustments. Despite some style adjustments, the functional changes are fixing the label-mac-device and removing the changes for RE650 v1, as that's a separate matter.

adrianschmutzler commented:
BTW: I'd be happy to update RE650v1 to DTS triggers as well, but only in a separate commit and only if tested.

achterin commented:
@gvlaev can you please test the trigger approach on you RE650?

gvlaev commented:
@achterin
@adrianschmutzler 

Yes, it works fine.

```
diff --git a/target/linux/ramips/dts/mt7621_tplink_re650-v1.dts b/target/linux/ramips/dts/mt7621_tplink_re650-v1.dts
index ab8805f163..2b1b6d609e 100644
--- a/target/linux/ramips/dts/mt7621_tplink_re650-v1.dts
+++ b/target/linux/ramips/dts/mt7621_tplink_re650-v1.dts
@@ -25,11 +25,13 @@
                wifi2g {
                        label = "re650-v1:blue:wifi2G";
                        gpios = <&gpio 23 GPIO_ACTIVE_LOW>;
+                       linux,default-trigger = "phy0tpt";
                };
 
                wifi5g {
                        label = "re650-v1:blue:wifi5G";
                        gpios = <&gpio 24 GPIO_ACTIVE_LOW>;
+                       linux,default-trigger = "phy1tpt";
                };
 
                wps_red {
diff --git a/target/linux/ramips/mt7621/base-files/etc/board.d/01_leds b/target/linux/ramips/mt7621/base-files/etc/board.d/01_leds
index 2e3dca7901..37ebd0fb64 100755
--- a/target/linux/ramips/mt7621/base-files/etc/board.d/01_leds
+++ b/target/linux/ramips/mt7621/base-files/etc/board.d/01_leds
@@ -68,14 +68,14 @@ netgear,r6800)
        ucidef_set_led_netdev "lan3" "LAN3" "$boardname:white:lan3" "lan3"
        ucidef_set_led_netdev "lan4" "LAN4" "$boardname:white:lan4" "lan4"
        ;;
-tplink,re350-v1|\
-tplink,re650-v1)
+tplink,re350-v1)
        ucidef_set_led_netdev "wifi2g" "Wifi 2.4G" "$boardname:blue:wifi2G" "wlan0"
        ucidef_set_led_netdev "wifi5g" "Wifi 5G" "$boardname:blue:wifi5G" "wlan1"
        ucidef_set_led_netdev "eth_act" "LAN act" "$boardname:green:eth_act" "lan" "tx rx"
        ucidef_set_led_netdev "eth_link" "LAN link" "$boardname:green:eth_link" "lan" "link"
        ;;
-tplink,re500-v1)
+tplink,re500-v1|\
+tplink,re650-v1)
        ucidef_set_led_netdev "eth_act" "LAN act" "$boardname:green:eth_act" "lan" "tx rx"
        ucidef_set_led_netdev "eth_link" "LAN link" "$boardname:green:eth_link" "lan" "link"
        ;;
```

adrianschmutzler commented:
@gvlaev Thanks, I've pushed a similar patch to master.
